### ISSUE TITLE
Develop a wrapper API for plotting properties from calculate and equilibrium

### ISSUE THREAD

**Issue by bocklund** [pycalphad/pycalphad#154]
An API to a general effect of `propplot(res_dataset, x='T', y='CPM')` (name TBD). Should work equally well for equilibrium properties and calculated properties and be generally aware of phases (e.g. different color line segments in equilibrium properties, and different lines in calculate properties)

This is somewhat parallel with #133 in that it would be nice to do one calculate call that you could call `propplot` on many times. 

For now, since it would be reasonable to plot many properties from one dataset, I think it makes sense that this has more of an `eqplot` signature (it takes a dataset) rather than a `binplot` or `ternplot` style signature.

**Comment by richardotis** [comment]
We could probably also use this to fix #55

**Comment by bocklund** [comment]
Some out of band discussion on this today. To summarize:

A new version of `eqplot` can handle plotting both phase diagrams and property diagrams. It should detect them automatically and dispatch to `pdplot` for phase diagrams and `propplot` for property diagrams. A version of a general version of `pdplot` for binary and ternary systems is written as the `eqplot` function [here](https://github.com/PhasesResearchLab/PSU-MatSE410/blob/master/ternary/eqplot_helper.py) (to be worked in to pycalphad). `pdplot` and `propplot` can exist as public "advanced usage" to allow for special cases when `eqplot` may make the wrong assumption.

The key distinction between `pdplot` and `propplot` is that it is assumed that `v.MoleFraction` (`v.X`) axes in `pdplot` should correspond to phase compositions (the way they `eq.X` is for an equilibrium `Dataset`) while `v.X` axes for `propplot` means overall composition (i.e. `np.dot(eq[X].sel(component='B'), eq[NP])` or `eq.X_B`).

It also needs to be decided how to give the axes for `Dataset` data variables that are not state variables (i.e. `eq.NP`). Are they just input as strings? For strings, maybe you try to look them up in xarray first, then fall back to getattr(pycalphad.variables, axis_var), then try to parse special cases v.X from the string.

In addition, the following features might be useful:
1. `propplot` can also take `Dataset` objects produced by `calculate` as well as `equilibrium`
2. `propplot` can automatically plot all the lines for coordinates (`Dataset.coords`) that are mapped over, but not the plot axes. For example, an equilibrium calculation with `T` and `X` independent compositions could be used to `propplot(eq, x=v.X('ZN'), y='GM')` plot all the lines across composition for different temperatures. Super cool if you could use this to loop over a custom user-added `parameters` coordinate.

**Comment by bocklund** [comment]
More OOB discussion on a proposed way to handle a property plotting API. 

As mapping APIs and CompositionSet-based APIs (e.g. #415) are getting closer to first class support, we can think about revisiting this one with `CompositionSet` objects as the basis.

Below, I sketched out a function that can compute properties from phase records using the DOF of the `CompositionSet` objects from a solution. The function allows for callers to override the `PhaseRecord` to be able to compute different properties, but needs to know the shape of the desired output and to provide a `PhaseRecord` for any property on-the-fly.

```python
def compset_sum(compsets: Sequence[PhaseRecord], output="obj", output_shape=(1,), transform=None, phase_records=None):
    # Example of a transform: suppose that output gives a 2D composition array
    # (output_shape=(2, 1)), but we only want the value of one composition
    if transform is None:
        # assume that, by default, phase_record.output produces a view of 1
        # dimension higher than the actual output of interest
        # That is, an energy is a 1D array that we are interested in the scalar
        # A composition is a 2D array that we are intersted in the 1D array... 
        _transform = lambda x: x[..., 0]
    else:
        _transform = transform

    phase_records = phase_records if phase_records is not None else {}

    total_value = _transform(np.zeros(output_shape))
    for compset in compsets:
        # Get the phase record if passed, otherwise fall back to the current composition set
        # This is primarily because our PhaseRecord objects can only compute one
        # output in the objective. So different outputs (i.e. "GM" vs. "HM")
        # are currently stored in different phase records.
        # By convention, a CompositionSet PhaseRecord will have the objective
        # be "GM" and it is safe to assume that any CompositionSet used for
        # minimization obeys that rule.
        phase_record = phase_records.get(compset.phase_record.phase_name, compset.phase_record)
        phase_record_method = getattr(phase_record, output)
        # call the method, being sure to pass the expected array buffer in as the first argument 
        output_arr = np.empty(output_shape, dtype="double")
        phase_record_method(output_arr, compset.dof)  # modified output_arr in place
        total_value += compset.NP * _transform(output_arr)
    return total_value
```

One requirement is to be able to compute properties on-the-fly imperatively. Currently, the `calculate` and `equilibrium` are declarative, where the properties must be given _before_ the calculation. 

One way to design this is to create a new `PropertyRecord` object (maybe one `PropertyRecord` for scalars and `CompositionPropertyRecord` for 1D properties of composition (e.g. moles)). `PropertyRecord` objects hold pointers to callaboes for the objective function, an (optional) gradient function, and an (optional) Hessian function (and similarly for the `CompositionPropertyRecord`). 

A `PhaseRecord` will then hold multiple `PropertyRecord`/`CompositionPropertyRecord` objects (which can be added to the `PhaseRecord` on-the-fly). Adding new `PropertyRecord` objects to a `PhaseRecord` on-the-fly may require use of the `Model` to compute the property symbolically and compile the callables to populate a `PropertyRecord`. From a design standpoint, it currently makes sense to have that live outside of `PhaseRecord` objects, so that `PhaseRecord` objects can continue to be independent of any symbolic information (i.e. `Model` objects).

To summarize:

- `PhaseRecord` objects hold many `PropertyRecord` objects
- `PropertyRecord` objects hold pointers to callables and are created from `Model` objects
